construct_runtime auto pallet parts: allow pallets to expand `{ Event, Call... }` automatically.

thiolliere opened the pull request:
### This PR is to be followed by its generalized usage in substrate https://github.com/paritytech/substrate/compare/gui-automatic-construct-runtime...gui-automatic-construct-runtime-usage?expand=1

### Description

Introduce new decl_construct_runtime_args so:

pallet declare construct_runtime args with: `decl_construct_runtime_args!( Module, Config<T>, Storage.. )`

and then construct_runtime can be called as such:

```rust
construct_runtime!(
  Treasury: frame_treasury,
)
```
this will expand to
```
construct_runtime!(
  Treasury: frame_treasury::{Module, Call, Storage, Config, Event<T>},
)
```

if there is multiple pallet in the same crate, a unique id must be provided, this unique id is only internally used:
`decl_construct_runtime_args!(#[unique_id = _2] Module, Config<T>, Storage.. )`

see https://github.com/paritytech/substrate/pull/6494/files#diff-a8500018e076e76e369b45df6010dd2d

### Goal

simplify construct_runtime so: less error prone construct_runtime and more usable in tests instead of impl_outer_* which are not consistent and very hacky.

### General notes:

* multiple pallet per crate is still doable

* no breaking change, runtime can still define the part themself

* This makes construct_runtime smaller, I think it should replace impl_outer_* in tests. Thus test environment are more consistent. (some helpful type has been added to system for this purpose)

* if you want to see how looks substrate once tests and runtime are moved to new syntax see https://github.com/paritytech/substrate/compare/gui-automatic-construct-runtime...gui-automatic-construct-runtime-usage?expand=1

* while implementing decl_construct_runtime_args for pallets I saw that some `Config` were missed.
  (maybe I could do more sanity check: if `Storage` is provided but not `Config` then decl_construct_runtime_args generate a test that `Config` is not generated by decl_storage (i.e. decl_storage will provide a function `__hidden_has_generated_config`). but this is for another PR and fixed by https://github.com/paritytech/substrate/issues/5678).

* https://github.com/paritytech/substrate/issues/5678 will implement this macro automatically.

### How does it works internally

for the given example
```rust
construct_runtime!(
  Treasury: frame_treasury,
  Treasury2: frame_treasury2,
)
```
construct_runtime do a preprocessing and expand for each `auto` pallet parts a overarching macro call:
```rust
frame_treasury::construct_runtime_args! {
	{ Treasury: frame_treasury } // mathing pattern
	frame_treasury2::construct_runtime_args! {
		{ Treasury2: frame_treasury2 } // matchine pattern
		construct_runtime! {
			Treasury: frame_treasury,
			Treasury2: frame_treasury2,
		}
	}
}
```
Each auto_construct_runtime replace the match of their matching pattern in the following token stream.
Resulting in:
```rust
frame_treasury2::construct_runtime_args! {
	{ Treasury2: frame_treasury2 } // matchine pattern
	construct_runtime! {
		Treasury: frame_treasury::{Module, Call, Storage, Config, Event<T>},
		Treasury2: frame_treasury2,
	}
}
```
until the expanded construct_runtime call:
```rust
construct_runtime! {
  Treasury: frame_treasury::{Module, Call, Storage, Config, Event<T>},
  Treasury2: frame_treasury2::{Module, Call, Storage, Config, Event<T>},
}
```

bkchr commented on frame/treasury/src/lib.rs:
```diff
@@ -111,6 +111,16 @@ type BalanceOf<T> = <<T as Trait>::Currency as Currency<<T as frame_system::Trai
 type NegativeImbalanceOf<T> = <<T as Trait>::Currency as Currency<<T as frame_system::Trait>::AccountId>>::NegativeImbalance;
 
+#[macro_export]
+macro_rules! auto_construct_runtime {
```
Nice idea, as you already said, it would be nice to replace this with
```
construct_runtime_args!( Module, Call, Storage, Config,  Event<T> )
```

`construct_runtime_args` can be replaced with a better name.

bkchr commented:
And we should not require `auto` keyword, it should be the default.

thiolliere commented:
Yes so final syntax should be something like:
```rust
construct_runtime!(.... { System: frame_system, Treasury: frame_treasury, OldSyntax: frame_old::{Module...} })
```
and for pallet side:
```rust
decl_construct_runtime_args!( Module, Call, Storage, Config,  Event<T> )
```

Though one thing is that construct_runtime_args doesn't really support `Inherent(Timestamp)` use by babe for instance. Though I think we can just not make use of this syntax anymore, and instead make babe make the check in on_finalize directly.

bkchr commented:
> Though one thing is that construct_runtime_args doesn't really support Inherent(Timestamp) use by babe for instance. Though I think we can just not make use of this syntax anymore, and instead make babe make the check in on_finalize directly.

I have something prepared for this already :D I need to revive the branch and open a Pr ;)

thiolliere commented:
I added better syntax and also now that construct_runtime is smaller I think we can use it in test instead of using impl_outer_*.
Test will consist of writing just a very small runtime: see pallet-assets in this PR

thiolliere commented:
I updated the PR to only contains the changes in frame-support frame-system and primitive, the generalized usage of it is moved to another branch https://github.com/paritytech/substrate/compare/gui-automatic-construct-runtime...gui-automatic-construct-runtime-usage?expand=1 where I implemented for all pallets, and most tests except non-obvious ones (i.e. system, balance, contract, and some other (just grep `impl_outer_` to see remainings usage))

thiolliere commented on primitives/runtime/src/lib.rs:
```diff
@@ -622,8 +622,13 @@ macro_rules! impl_outer_config {
 		$crate::paste::item! {
+			#[cfg(any(feature = "std", test))]
+			#[doc(hidden)]
+			use $crate::serde as _sp_runtime_hidden_serde;
```
This enforce only one call to impl_outer_config :-/ if this is considered a blocking point, I can try to work around by having construct_runtime giving the path to serde to impl_outer_config.

thiolliere replied:
Fixed

thiolliere commented:
The remaining question for me are:
* is it ok to have only one pallet define per crate:
  if not then we can think of this solution:
  
  allowing decl_construct_runtime_args to give an ident: (for instance pallet_session_historical), the args macro for the pallet would be construct_runtime_args_for_pallet_session_historal like:
  `decl_construct_runtime_args!(pallet_session_historical: Config, Event<T>, Storage)` it will define `construct_runtime_args_for_pallet_session_historical`

  then `construct_runtime!(..., MyModule: my_module, ..)` will call `my_module::construct_runtime_args_for_my_module`

* is it ok to have only one impl_outer_config per module (i.e. the macro can't be called two time in the same module). If not I can think of some stuff as well, but IMHO impl_outer_* macros shouldn't be used anymore once we have this more handy construct_runtime.

thiolliere commented:
All limitation have been solved:
* multiple pallet per crate is allowed
* no need for `local_macro` attribute in construct_runtime when using pallet define inside same crate

kianenigma commented on frame/support/procedural/src/construct_runtime/mod.rs:
```diff
@@ -163,6 +229,12 @@ fn decl_outer_inherent<'a>(
 	});
 	quote!(
+		// Prevent UncheckedExtrinsic to print unused warning.
+		const _: () = {
```
I see this pattern in the expansions of derive macros of codec as well, what is the purpose? is the keyword `__hidden_use_unchecked_extrinsic ` used anywhere now?

thiolliere replied:
not but this is a pattern used by may derive macro, for instance serde.
I think the main advantage is that it makes `type __hidden_use_unchecked_extrinsic` unavailable in doc and rust error will not recommand it.

kianenigma replied:
But if the type alias is not even accessible and used anywhere, I don't see how it is used. 

For example I've this and it makes sense: 

```rust
const _: () { impl Something for SomethingElse { ... } } 
``` 

but something like: 

```rust
const _: () = type __unused_alias = Foo; 
```

I don't get. 

Unless is `__hidden_use_unchecked_extrinsic` used in the other branch? Asking because I couldn't find any usage here.

thiolliere replied:
this is just to `// Prevent UncheckedExtrinsic to print unused warning.`
so I have `const _: () = { #[allow(unused)] type __unused_alias = Foo; }`
UncheckedExtrinsic is unused if there is no inherent.

kianenigma commented on frame/support/procedural/src/construct_runtime_args.rs:
```diff
@@ -0,0 +1,60 @@
+mod keyword {
```
license?

kianenigma commented on frame/support/procedural/src/expand_after.rs:
```diff
@@ -0,0 +1,115 @@
+use proc_macro2::{TokenStream, TokenTree, Group, Span};
```
license

kianenigma commented on frame/support/procedural/src/lib.rs:
````diff
@@ -280,12 +284,78 @@ pub fn decl_storage(input: TokenStream) -> TokenStream {
+/// will generate:
+///
+/// ```nocompile
+///     Some content with { at some point something or else } other match pattern are ignored
````
I don't get this example. Should it not replace `match pattern` with `expansion tokens`? 

Like: 

```suggestion
///     Some content with { at some point expansion tokens } other match pattern are ignored
```

thiolliere replied:
I forget to udpate this example, it should be:
```
    Some content with { at some point match pattern expansion tokens } other match pattern are ignored
```
thanks

kianenigma commented on frame/system/src/lib.rs:
```diff
@@ -1375,3 +1377,14 @@ impl<T: Trait> Lookup for ChainContext<T> {
 }
+
+/// An implementation of `sp_runtime::traits::Block` to be used in test.
+pub type MockBlock<T> = generic::Block<
```
This does not seem used to me, else why not feature gate it to test or at least std?

thiolliere replied:
in this following PR https://github.com/paritytech/substrate/compare/gui-automatic-construct-runtime...gui-automatic-construct-runtime-usage?expand=1
I used it a lot for pallet mocks.

I guess we could feature get but I don't see strong reason not to have in no_std as well. some people could make use of it for no_std mock.

kianenigma commented on frame/support/procedural/src/construct_runtime_args.rs:
```diff
@@ -0,0 +1,60 @@
+		/// This can be internally called by `construct_runtime` to builds the pallet args.
+		#[macro_export]
+		#[doc(hidden)]
+		macro_rules! #hidden_macro_name {
```
can you show point out where this is used? I couldn't find it anywhere.

thiolliere replied:
it is used in the preprocess of `construct_runtime!`
```rust
			expand = quote_spanned!(module_name.span() =>
				#module_module::construct_runtime_args! {
					{ #module_name : #module_module #module_instance}
					#expand
				}
			);
```

thiolliere commented:
One main issue of this PR is that pallet can remove their call or their origin and it changes the outer call and outer origin types.
maybe we should put it on ice until we have indices for pallet in construct_runtime.

gnunicorn commented:
closing because of inactivity.

thiolliere commented:
NOTE: I still think this makes definition of construct_runtime really easier. Either we should do like this or we could make new attribute macro implement an empty Event, Inherent, ValidateUnsigned, ... and make pallet in construct_runtime having by default all attribute (which is simpler).

bkchr commented:
> NOTE: I still think this makes definition of construct_runtime really easier. Either we should do like this or we could make new attribute macro implement an empty Event, Inherent, ValidateUnsigned, ... and make pallet in construct_runtime having by default all attribute (which is simpler).

Yeah, I'm also in for this.

One question, did your version supported negation? So let's say I have a pallet where I expose a `Call`, but I as a runtime developer don't want to expose this `Call`, could I write something like `!Call`?

thiolliere commented:
I didn't supported negation but indeed after macro attribute I can work on this

kianenigma commented:
Hope to see the PR be reincarnated soon.
